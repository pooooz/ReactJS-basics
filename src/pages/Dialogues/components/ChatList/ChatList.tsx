import React, { FC, useState } from 'react';
import { NavLink } from 'react-router-dom';
import { nanoid } from 'nanoid';

import { ChatInterface } from '../../../../App';

import styles from './ChatList.module.scss';

interface ChatListProps {
  chatList: ChatInterface[];
  onAddChat: (chats: ChatInterface) => void;
  onDeleteChat: (chatName: string) => void;
}

export const ChatList: FC<ChatListProps> = (props) => {
  const [name, setName] = useState('');

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (name) {
      props.onAddChat({
        id: nanoid(),
        name,
      });
      setName('');
    }
  };

  return (
    <aside className={styles.chatList}>
      <div className={styles.chatHeader}>Chats</div>
      <ul>
        {props.chatList.map((chat) => (
          <li className={styles.chatItem} key={chat.id}>
            <NavLink to={`/chats/${chat.name}`} className={styles.link}>
              {chat.name}
            </NavLink>
            <button
              className={styles.deleteButton}
              onClick={() => props.onDeleteChat(chat.name)}
            >
              X
            </button>
          </li>
        ))}
      </ul>
      <form onSubmit={handleSubmit} className={styles.form}>
        <div className={styles.inputWrap}>
          <input
            type="text"
            className={styles.input}
            value={name}
            onChange={(e) => setName(e.target.value)}
          />
        </div>
        <button className={styles.button} type="submit">
          Add chat
        </button>
      </form>
    </aside>
  );
};
