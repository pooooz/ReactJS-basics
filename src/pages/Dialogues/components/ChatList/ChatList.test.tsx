import React from 'react';
import '@testing-library/jest-dom';
import '@testing-library/user-event';
import { fireEvent, render, screen } from '@testing-library/react';
import { ChatList } from './ChatList';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

describe('ChatList test', () => {
  const onAddChat = jest.fn();
  const onDeleteChat = jest.fn();
  it('Render component', () => {
    const history = createMemoryHistory();
    render(
      <Router location="/chats" navigator={history}>
        <ChatList
          chatList={[{ id: '1', name: 'initial' }]}
          onAddChat={onAddChat}
          onDeleteChat={onDeleteChat}
        />
      </Router>
    );
  });

  it('Adding a chat', () => {
    const history = createMemoryHistory();
    render(
      <Router location="/chats" navigator={history}>
        <ChatList
          chatList={[{ id: '1', name: 'initial' }]}
          onAddChat={onAddChat}
          onDeleteChat={onDeleteChat}
        />
      </Router>
    );
    fireEvent.input(screen.getByRole('textbox'), {
      target: { value: 'myChat' },
    });
    fireEvent.click(screen.getByRole('button', { name: /Add chat/ }));
    expect(screen.getByText(/initial/)).toBeInTheDocument();
  });

  it('Deleting a chat', () => {
    const history = createMemoryHistory();
    render(
      <Router location="/chats" navigator={history}>
        <ChatList
          chatList={[{ id: '1', name: 'initial' }]}
          onAddChat={onAddChat}
          onDeleteChat={onDeleteChat}
        />
      </Router>
    );
    fireEvent.click(screen.getByRole('button', { name: /X/ }));
  });

  it('Adding an empty chat', () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Router location="/chats" navigator={history}>
        <ChatList
          chatList={[{ id: '1', name: 'initial' }]}
          onAddChat={onAddChat}
          onDeleteChat={onDeleteChat}
        />
      </Router>
    );
    fireEvent.input(screen.getByRole('textbox'), {
      target: { value: '' },
    });
    fireEvent.click(screen.getByRole('button', { name: /Add chat/ }));
    expect(asFragment()).toMatchSnapshot();
  });
});
