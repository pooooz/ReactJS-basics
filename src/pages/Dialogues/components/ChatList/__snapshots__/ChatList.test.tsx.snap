// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatList test Adding an empty chat 1`] = `
<DocumentFragment>
  <aside>
    <div>
      Chats
    </div>
    <ul>
      <li>
        <a
          class=""
          href="/chats/initial"
        >
          initial
        </a>
        <button>
          X
        </button>
      </li>
    </ul>
    <form>
      <div>
        <input
          type="text"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        Add chat
      </button>
    </form>
  </aside>
</DocumentFragment>
`;
